### PR TITLE
Integer K-Means and Hierarchical K-Means

### PR DESCRIPTION
**Pull request by Atanahel**
Because I was started so why not finish...

Was a great way to learn more about Cython, especially the Hierarchical part which is done by converting to C struct to a python representation back and forth (as it is done in the MATLAB version).

Test cases are trickier to design because of the random initialization of the algorithm which does not always give correct results for toy datasets.

### CODE REVIEW

**Comment by Atanahel** [comment]
I don't understand the error AppVeyor is getting... @patricksnape can you rerun the checks?

**Comment by patricksnape** [comment]
Jenkins test this please

**Comment by patricksnape** [comment]
@Atanahel I think its segfaulting on Python 3.5 on Windows

**Comment by Atanahel** [comment]
Seems so, but I'm very confused at why it'd work on the different architectures on python 3.5 and with windows on python 3.4 but **not** on windows+py3.5...

**Comment by patricksnape** [comment]
It uses a different compiler unfortunately, so that is likely why.

**Comment by patricksnape** [comment]
Can you just merge master into here so we can see if this still fails? If so I will take a look at it!

**Comment by Atanahel** [comment]
Youhou, passing!

No clues why it changed though...

**Comment by patricksnape** [comment]
I'm going to put my hand up and say that I think the vlfeat builds on Windows were actually broken - and I fixed them when I went to enable threads. The wrong runtime was being linked to which was causing memory issues. So thanks for forcing me to notice that! I will review this PR asap :+1:

File: cyvlfeat/kmeans/ikmeans.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+    if (not isinstance(max_num_iterations, int)) or max_num_iterations <= 0:
+        raise ValueError('max_num_iterations should be a positive integer')
+
+    return cy_ikmeans(data, num_centers, algorithm_b, max_num_iterations, verbose)
+
+
+def ikmeans_push(data, centers):
+    """
+
```
Comment (patricksnape) [inline cyvlfeat/kmeans/ikmeans.py]: What does this method do?

File: cyvlfeat/kmeans/kmeans.py [hunk 2]
```diff
@@ -82,7 +83,73 @@ def kmeans(data, num_centers, distance="l2", initialization="RANDSEL",
...
     return cy_kmeans(data, num_centers, distance_b, initialization_b,
                      algorithm_b, num_repetitions, num_trees, max_num_comparisons,
                      max_num_iterations, min_energy_variation, verbose)
+
+
+def kmeans_quantize(data, centers, distance="l2", algorithm="LLOYD", num_trees=3,
+                    max_num_comparisons=100, verbose=False):
+    """
+    Project the given data to the corresponding cluster indices.
+    Parameters
```
Comment (patricksnape) [inline cyvlfeat/kmeans/kmeans.py]: Newline required

**Comment by patricksnape** [comment]
Mostly just not sure what that push method does - if you just update the docs for this I will merge and make a new release :smile:

**Comment by Atanahel** [comment]
Haha, yeah I kept the terminology of the Matlab interface but I agree that this name is not self-explanatory at all.

It's actually a method that from a set of pre-computed centers, compute the nearest neighbour center for each point in a set of given points. It's for projecting the visual words on the dictionnary of the k-means centers.

**Comment by patricksnape** [comment]
Great! Thanks so much for the excellent contributions! Merging now since we know it passes :smile:

**Comment by Atanahel** [comment]
Cool!

By the way, maybe the documentation of the starting page should be updated. It still says only the SIFT methods are interfaced which is not true anymore (HoG, FV, K-means now)

**Comment by patricksnape** [comment]
Updated!

**Pull request merged**